src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/ResourceScope.java

```diff
@@ -0,0 +1,289 @@
+ */
+public interface ResourceScope extends AutoCloseable {
+    /**
+     * Is this resource scope alive?
+     * @return true, if this resource scope is alive.
+     * @see ResourceScope#close()
+     */
+    boolean isAlive();
+
+    /**
+     * The thread owning this resource scope.
+     * @return the thread owning this resource scope, or {@code null} if this resource scope is shared.
+     */
+    Thread ownerThread();
+
+    /**
+     * Is this resource scope an <em>implicit scope</em>?
+     * @return true if this scope is an <em>implicit scope</em>.
+     * @see #newImplicitScope()
+     * @see #globalScope()
+     */
+    boolean isImplicit();
+
+    /**
+     * Closes this resource scope. As a side-effect, if this operation completes without exceptions, this scope will be marked
+     * as <em>not alive</em>, and subsequent operations on resources associated with this scope will fail with {@link IllegalStateException}.
+     * Additionally, upon successful closure, all native resources associated with this resource scope will be released.
+     *
+     * @apiNote This operation is not idempotent; that is, closing an already closed resource scope <em>always</em> results in an
+     * exception being thrown. This reflects a deliberate design choice: resource scope state transitions should be
+     * manifest in the client code; a failure in any of these transitions reveals a bug in the underlying application
+     * logic.
+     *
+     * @throws IllegalStateException if one of the following condition is met:
+     * <ul>
+     *     <li>this resource scope is not <em>alive</em>
+     *     <li>this resource scope is confined, and this method is called from a thread other than the thread owning this resource scope</li>
+     *     <li>this resource scope is shared and a resource associated with this scope is accessed while this method is called</li>
+     *     <li>one or more handles (see {@link #acquire()}) associated with this resource scope have not been closed</li>
+     * </ul>
+     * @throws UnsupportedOperationException if this resource scope is {@link #isImplicit() implicit}.
+     */
+    void close();
+
+    /**
+     * Add a custom cleanup action which will be executed when the resource scope is closed.
+     * The order in which custom cleanup actions are invoked once the scope is closed is unspecified.
+     * @param runnable the custom cleanup action to be associated with this scope.
+     * @throws IllegalStateException if this scope has already been closed.
+     */
+    void addCloseAction(Runnable runnable);
+
+    /**
+     * If this resource scope is explicit, this method acquires a new resource scope handle, associated with this
+     * resource scope; an explicit resource scope cannot be {@link #close() closed}
+     * until all the resource scope handles acquired from it have been {@link #release(Handle)} released}.
+     * @return a resource scope handle.
+     */
```

Comment:
Given recent changes, it might be good to generalise the opening sentence here. Maybe :
 " Acquires a resource scope handle associated with this resource scope. If the resource scope is explicit ... "   Or some such.

Reply:
I'm afraid I don't get this comment ;-)

Reply:
The spec for the acquire method talks only of explicit resource scopes. The comment is suggesting that the spec could be generalised a little, since it is possible to acquire a resource scope handle associated with implicit scopes.